Replace pytree_assert with production pytree_check. Remove pytree_unreachable

swolchok opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #7655
* #7654
* #7653

When handling untrusted input, it's not appropriate to use debug-only checks; we should be checking in prod as these are not programmer errors. pytree_unreachable was similarly being used for input validation.

Differential Revision: [D68166301](https://our.internmc.facebook.com/intern/diff/D68166301/)

Commit:
Replace pytree_assert with production pytree_check. Remove pytree_unreachable

When handling untrusted input, it's not appropriate to use debug-only checks; we should be checking in prod as these are not programmer errors. pytree_unreachable was similarly being used for input validation.

Differential Revision: [D68166301](https://our.internmc.facebook.com/intern/diff/D68166301/)

[ghstack-poisoned]

swolchok committed:
Update on "Replace pytree_assert with production pytree_check. Remove pytree_unreachable"


When handling untrusted input, it's not appropriate to use debug-only checks; we should be checking in prod as these are not programmer errors. pytree_unreachable was similarly being used for input validation.

Differential Revision: [D68166301](https://our.internmc.facebook.com/intern/diff/D68166301/)

[ghstack-poisoned]

swolchok committed:
Update on "Replace pytree_assert with production pytree_check. Remove pytree_unreachable"


When handling untrusted input, it's not appropriate to use debug-only checks; we should be checking in prod as these are not programmer errors. pytree_unreachable was similarly being used for input validation.

Differential Revision: [D68166301](https://our.internmc.facebook.com/intern/diff/D68166301/)

[ghstack-poisoned]

dbort commented on extension/pytree/pybindings.cpp:
```diff
@@ -145,7 +145,8 @@ class PyTree {
           } else {
-            pytree_assert(false);
+            throw std::runtime_err(
+                "invalid key in pytree dict; must be int or string");
```
It'd be helpful to print the key type here, though the old code didn't do it either so not a blocker. Similar for other new `throw` messages in this PR.

dbort reviewed (approved):
Much nicer behavior than before!

swolchok committed:
Update on "Replace pytree_assert with production pytree_check. Remove pytree_unreachable"


When handling untrusted input, it's not appropriate to use debug-only checks; we should be checking in prod as these are not programmer errors. pytree_unreachable was similarly being used for input validation.

Differential Revision: [D68166301](https://our.internmc.facebook.com/intern/diff/D68166301/)

[ghstack-poisoned]

swolchok committed:
Update on "Replace pytree_assert with production pytree_check. Remove pytree_unreachable"


When handling untrusted input, it's not appropriate to use debug-only checks; we should be checking in prod as these are not programmer errors. pytree_unreachable was similarly being used for input validation.

Differential Revision: [D68166301](https://our.internmc.facebook.com/intern/diff/D68166301/)

[ghstack-poisoned]

Commit:
Update on "Replace pytree_assert with production pytree_check. Remove pytree_unreachable"


When handling untrusted input, it's not appropriate to use debug-only checks; we should be checking in prod as these are not programmer errors. pytree_unreachable was similarly being used for input validation.

Differential Revision: [D68166301](https://our.internmc.facebook.com/intern/diff/D68166301/)

[ghstack-poisoned]
